jpa/subsystem/src/main/java/org/jboss/as/jpa/container/TransactionScopedEntityManager.java

```diff
@@ -198,6 +198,11 @@ protected boolean skipQueryDetach() {
         return skipQueryDetach.booleanValue();
     }
 
+    @Override
+    public void setProperty(String propertyName, Object value) {
+        properties.put(propertyName, value);
```

Comment:
I found at least one call path (via PersistenceUnitServiceHandler) that leads to 'properties' being assigned to an immutable map, which would cause this to fail. There may be others.

Thread safety is probably an issue as well since AIUI this is part of the EM API and thus can be called from application.

Reply:
The EntityManager is only expected to be accessed by one application thread at a time but we should ensure that the property added to the HashMap is not left in thread memory:

For example:

1. Servlet request makes call to session bean that adds calls EntityManager.setProperty("myAppProperty", new MyAppClassPropertyValue).  The session bean call happens to run on thread id = 2001.
2. Servlet request than calls the session bean again and the session bean happens to run on different thread id = 2025.  

The HashMap itself should be available on all threads once it is created but I think that values put in the TransactionScopedEntityManager as properties may be cached in CPU registers.

Reply:
More about thread safety:

As per https://jakarta.ee/specifications/persistence/3.2/jakarta-persistence-spec-3.2#obtaining-an-entitymanager

> An entity manager must not be shared among multiple concurrently executing threads, as the entity manager and persistence context are not required to be threadsafe. Entity managers must only be accessed in a single-threaded manner.